minetest/minetest
# TouchScreenGUI dehardcoding refactor

Some TouchScreenGUI refactoring is required to make customizable touchscreen controls (#14456) possible. This PR changes the following things from being hardcoded at compile time to being determined at runtime:

- which / how many buttons are on the main screen
- which / how many buttons are inside buttonbars
- which / how many buttonbars exist

Other things achieved by this PR:

- elimination of behavior differences between buttons on the main screen and in buttonbars
- less code duplication
- removal of separate "initialize" functions -> no "uninitialized" state causing crashes like #13522
- `std::shared_ptr` with a custom deleter (`irr_ptr` doesn't work with GUI elements) instead of manual refcounting
- overall ~200 lines of code less (as of 03/17/2024)

buttonbar = a row of buttons that can be expanded and collapsed

## To do

This PR is a Ready for Review.

Hiding whitespace changes is recommend.

## How to test

Beginner: Verify that the touch controls still work.

Intermediate: Verify that the dehardcoding was successful by changing some stuff in the TouchScreenGUI constructor.
Example: [diff.txt](https://github.com/minetest/minetest/files/14628603/diff.txt)
